src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -410,14 +412,10 @@ private async ValueTask<FrameHeader> ReadFrameAsync(bool initialFrame = false)
                     _incomingBuffer.Commit(bytesRead);
                     if (bytesRead == 0)
                     {
-                        if (_incomingBuffer.ActiveLength == 0)
-                        {
-                            ThrowMissingFrame();
-                        }
-                        else
-                        {
-                            ThrowPrematureEOF(FrameHeader.Size);
-                        }
+                        HttpIOException exception = _incomingBuffer.ActiveLength == 0 ?
+                            GetMissingFrameException() : GetPrematureEOFException(FrameHeader.Size);
+                        throw _goAwayErrorCode is not null ?
+                            new HttpProtocolException((long)_goAwayErrorCode, exception.Message, exception) : exception;
```

Comment:
```suggestion
                        if (_goAwayErrorCode is not null)
                        {
                            ThrowProtocolError(_goAwayErrorCode.Value, SR.net_http_http2_connection_close);
                        }
                        else if (_incomingBuffer.ActiveLength == 0)
                        {
                            ThrowMissingFrame();
                        }
                        else
                        {
                            ThrowPrematureEOF(FrameHeader.Size);
                        }
```

I think we should match the exception message you'd get from a GoAway normally here.